pkg/mimirpb/mimir.proto

```diff
@@ -169,8 +170,9 @@ message BucketSpan {
 }
 
 message FloatHistogramPair {
-  FloatHistogram histogram = 1 [(gogoproto.nullable) = false];
-  int64 timestamp_ms = 2;
+	// Fields order MUST match promql.HPoint so that we can cast types between them.
+  int64 timestamp_ms = 1;
+  FloatHistogram histogram = 2 [(gogoproto.nullable) = true];
```

Comment:
Also, it would be good to benchmark the performance impact of making the `histogram` field nullable (and therefore a pointer) and being able to directly cast vs leaving it non-nullable but not being able to directly cast between types.

Reply:
The 0ns results for `FromFPointsToSamples` and `FromHPointsToHistograms` seem suspicious - could you please check that the calls to `FromFPointsToSamples` and `FromHPointsToHistograms` aren't being optimised away? (Perhaps we need to store the result from each iteration?)

Reply:
Updated benchmark results between `5081ace` and `d0a8045`:
```
name                                   old time/op    new time/op    delta
FromFPointsToSamples-16                   302ns ± 2%       0ns ± 1%   -99.92%  (p=0.000 n=10+9)
FromHPointsToHistograms-16               4.61µs ± 3%    0.00µs ± 1%  -100.00%  (p=0.000 n=9+10)
FromHistogramToHistogramProto-16         21.7µs ± 2%    11.6µs ± 1%   -46.72%  (p=0.000 n=10+9)
FromFloatHistogramToHistogramProto-16    21.5µs ± 2%    11.5µs ± 1%   -46.31%  (p=0.000 n=10+9)

name                                   old alloc/op   new alloc/op   delta
FromFPointsToSamples-16                  1.79kB ± 0%    0.00kB       -100.00%  (p=0.000 n=10+10)
FromHPointsToHistograms-16               16.4kB ± 0%     0.0kB       -100.00%  (p=0.000 n=10+10)
FromHistogramToHistogramProto-16         22.4kB ± 0%    16.0kB ± 0%   -28.57%  (p=0.000 n=10+10)
FromFloatHistogramToHistogramProto-16    22.4kB ± 0%    16.0kB ± 0%   -28.57%  (p=0.000 n=10+10)

name                                   old allocs/op  new allocs/op  delta
FromFPointsToSamples-16                    1.00 ± 0%      0.00       -100.00%  (p=0.000 n=10+10)
FromHPointsToHistograms-16                 1.00 ± 0%      0.00       -100.00%  (p=0.000 n=10+10)
FromHistogramToHistogramProto-16            700 ± 0%       300 ± 0%   -57.14%  (p=0.000 n=10+10)
FromFloatHistogramToHistogramProto-16       700 ± 0%       300 ± 0%   -57.14%  (p=0.000 n=10+10)
```

Storing the result from each iteration doesn't seem to make a difference. Looks like it's just rounding off, I've attached the results from running the benchmark before and after.
[Benchmark_before.txt](https://github.com/grafana/mimir/files/11332676/Benchmark_before.txt)
[Benchmark_after.txt](https://github.com/grafana/mimir/files/11332677/Benchmark_after.txt)